Codec/Archive/Tar/Types.hs

```diff
@@ -462,46 +467,48 @@ newtype LinkTarget = LinkTarget BS.ByteString
 instance NFData LinkTarget where
     rnf (LinkTarget bs) = rnf bs
 
--- | Convert a native 'FilePath' to a tar 'LinkTarget'. This may fail if the
+-- | Convert a native 'FilePath' to a tar 'LinkTarget'.
 -- string is longer than 100 characters or if it contains non-portable
 -- characters.
---
-toLinkTarget   :: FilePath -> Maybe LinkTarget
-toLinkTarget path | length path <= 100 = Just $! LinkTarget (packAscii path)
-                  | otherwise          = Nothing
+toLinkTarget   :: MonadThrow m => FilePath -> m LinkTarget
+toLinkTarget path | length path <= 100 = do
+                                           target <- toLinkTarget' path
+                                           pure $! LinkTarget (packAscii target)
+                  | otherwise          = throwM (TooLong path)
+
+data LinkTargetException = IsAbsolute FilePath
+                         | TooLong FilePath
+  deriving (Show,Typeable)
+
+instance Exception LinkTargetException where
+  displayException (IsAbsolute fp) = "Link target \"" <> fp <> "\" is unexpectedly absolute"
+  displayException (TooLong _) = "The link target is too long"
+
+-- | Convert a native 'FilePath' to a unix filepath suitable for
+-- using as 'LinkTarget'. Does not error if longer than 100 characters.
+toLinkTarget' :: MonadThrow m => FilePath -> m FilePath
```

Comment:
If this function can throw in exactly one way, I'd prefer to return `Maybe`:
```suggestion
toLinkTarget' :: FilePath -> Maybe FilePath
```
...and we'd avoid depending on `exceptions` (which is not a big deal, but still nice).

Reply:
But `toLinkTarget` can now throw two different constructors. I can convert to `Either`, but I like MonadThrow, because it converts into IO exception automatically.  Not a strong opinion though.

Reply:
I'd rather be very explicit and monomorphic, thus `Either` is better. But not a strong opinion either :)